module2/gift-certificate-system/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java

```diff
@@ -0,0 +1,151 @@
+                                                                 UriComponentsBuilder ucBuilder) {
+
+        try {
+            GiftCertificate oldCertificate = giftCertificateService.findCertificateById(id);
+            GiftCertificate certificatePatched = applyPatchToGiftCertificate(patch, oldCertificate);
+
+            BindingResult result = new BeanPropertyBindingResult(certificatePatched, "giftCertificate");
+            certificateValidator.validate(certificatePatched, result);
+            if (result.hasErrors()) {
+                return new ResponseEntity(result.getAllErrors(), HttpStatus.BAD_REQUEST);
+            } else {
+                giftCertificateService.updateCertificate(certificatePatched);
+                HttpHeaders headers = new HttpHeaders();
+                headers.setLocation(ucBuilder.path("/certificates/{id}").buildAndExpand(certificatePatched.getId()).toUri());
+                return new ResponseEntity<>(headers, HttpStatus.CREATED);
+            }
+        } catch (JsonPatchException | JsonProcessingException e) {
+            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
+        } catch (GiftCertificateNotFoundException | TagNotFoundException e) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
+        }
+    }
+
+    @DeleteMapping(path = "certificates/{id}")
+    public ResponseEntity<Void> deleteGiftCertificate(@PathVariable Long id) {
+
+        giftCertificateService.deleteCertificate(id);
+        return new ResponseEntity<>(HttpStatus.OK);
+    }
+
+
+    @GetMapping(value = "certificates/{id}")
+    public GiftCertificate findCertificateById(@PathVariable Long id) {
+        return giftCertificateService.findCertificateById(id);
+    }
+
+    @GetMapping(value = "/certificates")
+    public List<GiftCertificate> findAll(@RequestParam(value = "sort_name") Optional<String> nameSortDirection,
+                                         @RequestParam(value = "sort_create_date") Optional<String> createDateDirection) {
+        List<GiftCertificate> certificates = giftCertificateService.findAllCertificates();
+        if (nameSortDirection.isPresent()) {
+            return giftCertificateService.sortCertificateByParameters("sort_name", nameSortDirection.get(), certificates);
+        } else if (createDateDirection.isPresent()) {
+            return giftCertificateService.sortCertificateByParameters("sort_create_date", createDateDirection.get(), certificates);
+        } else {
+            return certificates;
+        }
+    }
+
+    @GetMapping(value = "/search")
+    public List<GiftCertificate> findCertificateByParameters(@RequestParam(value = "tag_name") Optional<String> tagName,
+                                                             @RequestParam(value = "part_of_name") Optional<String> partOfName,
+                                                             @RequestParam(value = "part_of_description") Optional<String> partOfDescription) {
+
```

Comment:
As option, create Query object that contains all filtering options and send it to the service method as parameter, for example
```
Query query = new Query();
query.setTagFilter(tag);
query.setNameFilter(name);

giftCertificateService.getCeriticates(query);
```